File: BFS/Binary Tree Right Side View/Binary Tree Right Side View.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
      List<Integer> res = new ArrayList<>();   
      Queue<TreeNode> q = new LinkedList<>();
      if(root == null) return res;
      q.offer(root);
      while(!q.isEmpty()){
          int n = q.size();
          for(int i = 0; i < n; i++){
            TreeNode curr = q.poll();
          // when i == 0 that mean we have the right most node in that level.
              if( i == 0 ) res.add(curr.val);
              if(curr.right != null ) q.offer(curr.right);
              if(curr.left != null ) q.offer(curr.left);
          }
      }
        return res;
    }
}

File: BFS/Binary Tree Right Side View/Binary Tree Right Side View.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
       vector<int> res;
        queue<TreeNode*> q;
        if(root == nullptr) return res;
        q.push(root);
        while( !q.empty() ){
            int n = q.size();
            TreeNode* right = nullptr;
            for(int i = 0; i < n; i++){
               //cout<<"q.pop() : "<< q.pop() << endl;
                TreeNode* curr = q.front();
                q.pop();
              if(i == 0) res.push_back(curr->val);
              if(curr->right != nullptr) q.push(curr->right);
              if(curr->left != nullptr) q.push(curr->left);
            }
        }
        return res;
    }
};
